js/impress.js

```diff
@@ -659,6 +705,64 @@
         return lib;
     };
 
+    // `addPreInitPlugin` allows plugins to register a function that should
+    // be run (synchronously) at the beginning of init, before
+    // impress().init() itself executes.
+    impress.addPreInitPlugin = function( plugin, weight ) {
+        weight = parseInt( weight ) || 10;
+        if ( weight <= 0 ) {
+            throw "addPreInitPlugin: weight must be a positive integer";
+        }
+
+        if ( preInitPlugins[ weight ] === undefined ) {
+            preInitPlugins[ weight ] = [];
+        }
+        preInitPlugins[ weight ].push( plugin );
+    };
+
+    // Called at beginning of init, to execute all pre-init plugins.
+    var execPreInitPlugins = function( root ) { //jshint ignore:line
+        for ( var i = 0; i < preInitPlugins.length; i++ ) {
+            var thisLevel = preInitPlugins[ i ];
```

Comment:
What's gonna happen when two pre-init plugins have the same `weight`?

Reply:
The official answer is undefined. You should use different weights when 2 plugins have some inter-dependency, to ensure correct ordering.

They are pushed into an array and executed on a FIFO basis. So in practice the order of plugins in build.js determines the order when weight is equal.

Reply:
What about scalability? Right now the building of plugins is centralized, so it's easy to keep track of the `weight` numbers. If we start exposing to the public it's unlikely that plugin authors will be able to keep track of the ordering.

Is there a way plugins can be created so that they don't have inter-dependencies?